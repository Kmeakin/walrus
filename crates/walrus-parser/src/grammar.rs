use nom::Parser;
use nom_supreme::parser_ext::ParserExt;
use walrus_lexer::Token;
use walrus_syntax::{nodes::*, tokens::*};

mod support;
mod tokens;

mod decl;
mod expr;
mod lit;
mod pat;
mod ty;

pub use self::{decl::*, expr::*, lit::*, pat::*, support::*, tokens::*, ty::*};

#[cfg(test)]
use crate::test_parse;

pub type Err = ();
pub type IResult<'i, T> = nom::IResult<Input<'i>, T, Err>;
pub type Input<'i> = &'i [Token<'i>];

#[cfg(test)]
#[macro_export]
macro_rules! test_parse {
    ($name:ident, $parser:expr, $src:expr) => {
        #[test]
        fn $name() { test_parse($parser, $src) }
    };
}

#[cfg(test)]
fn test_parse<T>(mut parser: impl Fn(Input) -> IResult<T>, src: &'static str)
where
    T: std::fmt::Debug,
{
    use insta::*;

    let tokens = walrus_lexer::lex(src)
        .filter(|token| !token.kind.is_trivia())
        .collect::<Vec<_>>()
        .clone();
    let got = parser.parse(&tokens);

    let mut settings = insta::Settings::new();
    settings.set_snapshot_path("../snapshots");
    settings.set_prepend_module_to_snapshot(false);
    settings.bind(|| assert_debug_snapshot!(got));
}
