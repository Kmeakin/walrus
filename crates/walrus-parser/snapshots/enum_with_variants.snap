---
source: crates/walrus-parser/src/grammar.rs
expression: got

---
Ok(
    (
        [],
        SourceFile {
            decls: [
                Enum(
                    EnumDef {
                        kw_enum: KwEnum {
                            span: 0..4,
                        },
                        name: Var(
                            Ident {
                                span: 5..14,
                                text: "OptionInt",
                            },
                        ),
                        variants: Delimited {
                            open: LCurly {
                                span: 15..16,
                            },
                            inner: Punctuated0 {
                                first: Some(
                                    EnumVariant {
                                        name: Var(
                                            Ident {
                                                span: 29..33,
                                                text: "None",
                                            },
                                        ),
                                        fields: Delimited {
                                            open: LCurly {
                                                span: 34..35,
                                            },
                                            inner: Punctuated0 {
                                                first: None,
                                                tail: [],
                                                trail: None,
                                            },
                                            close: RCurly {
                                                span: 35..36,
                                            },
                                        },
                                    },
                                ),
                                tail: [
                                    (
                                        Comma {
                                            span: 36..37,
                                        },
                                        EnumVariant {
                                            name: Var(
                                                Ident {
                                                    span: 50..54,
                                                    text: "Some",
                                                },
                                            ),
                                            fields: Delimited {
                                                open: LCurly {
                                                    span: 55..56,
                                                },
                                                inner: Punctuated0 {
                                                    first: Some(
                                                        StructField {
                                                            name: Var(
                                                                Ident {
                                                                    span: 57..60,
                                                                    text: "val",
                                                                },
                                                            ),
                                                            colon: Colon {
                                                                span: 60..61,
                                                            },
                                                            ty: Var(
                                                                Var(
                                                                    Ident {
                                                                        span: 62..65,
                                                                        text: "Int",
                                                                    },
                                                                ),
                                                            ),
                                                        },
                                                    ),
                                                    tail: [],
                                                    trail: None,
                                                },
                                                close: RCurly {
                                                    span: 66..67,
                                                },
                                            },
                                        },
                                    ),
                                ],
                                trail: None,
                            },
                            close: RCurly {
                                span: 76..77,
                            },
                        },
                    },
                ),
            ],
        },
    ),
)
